google-cloud-pubsub/test/google/cloud/pubsub/flow_controller_test.rb

```diff
@@ -0,0 +1,545 @@
+      flow_controller = Google::Cloud::PubSub::FlowController.new(
+        message_limit: 1000,
+        byte_limit: 3,
+        limit_exceeded_behavior: :block
+      )
+
+      adding_1_done = Concurrent::Event.new
+      adding_2_done = Concurrent::Event.new
+      adding_3_done = Concurrent::Event.new
+      releasing_1_done = Concurrent::Event.new
+      releasing_2_done = Concurrent::Event.new
+      releasing_3_done = Concurrent::Event.new
+
+      _(flow_controller.outstanding_bytes).must_equal 0 # Implementation detail
+
+      run_in_thread flow_controller, :acquire, 3, adding_1_done
+      assert adding_1_done.wait(0.1), "Adding message 1 never unblocked."
+      _(flow_controller.outstanding_bytes).must_equal 3 # Implementation detail
+
+      run_in_thread flow_controller, :acquire, 3, adding_2_done
+      refute adding_2_done.wait(0.1), "Adding message 2 did not block."
+
+      run_in_thread flow_controller, :acquire, 3, adding_3_done
+      refute adding_3_done.wait(0.1), "Adding message 3 did not block."
+
+      run_in_thread flow_controller, :release, 3, releasing_1_done
+      assert releasing_1_done.wait(0.1), "Releasing message 1 errored."
+
+      assert adding_2_done.wait(0.1), "Adding message 2 never unblocked."
+
+      run_in_thread flow_controller, :release, 3, releasing_2_done
+      assert releasing_2_done.wait(0.1), "Releasing message 2 errored."
+
+      assert adding_3_done.wait(0.1), "Adding message 3 never unblocked."
+      _(flow_controller.outstanding_bytes).must_equal 3 # Implementation detail
+
+      run_in_thread flow_controller, :release, 3, releasing_3_done
+      assert releasing_3_done.wait(0.1), "Releasing message 3 errored."
+      _(flow_controller.outstanding_bytes).must_equal 0 # Implementation detail
+    end
+
+    it "unblocks all waiting acquires to byte_limit after a single release of sufficient bytes" do
+      flow_controller = Google::Cloud::PubSub::FlowController.new(
+        message_limit: 1000,
+        byte_limit: 3,
+        limit_exceeded_behavior: :block
+      )
+
+      adding_1_done = Concurrent::Event.new
+      adding_2_done = Concurrent::Event.new
+      adding_3_done = Concurrent::Event.new
+      adding_4_done = Concurrent::Event.new
+      releasing_1_done = Concurrent::Event.new
+
+      run_in_thread flow_controller, :acquire, 3, adding_1_done
+      assert adding_1_done.wait(0.1), "Adding message 1 never unblocked."
+      _(flow_controller.outstanding_bytes).must_equal 3 # Implementation detail
+
+      run_in_thread flow_controller, :acquire, 1, adding_2_done
+      refute adding_2_done.wait(0.1), "Adding message 2 did not block."
+
+      run_in_thread flow_controller, :acquire, 1, adding_3_done
+      refute adding_3_done.wait(0.1), "Adding message 3 did not block."
+
+      run_in_thread flow_controller, :acquire, 1, adding_4_done
+      refute adding_4_done.wait(0.1), "Adding message 4 did not block."
+
+      flow_controller.release 40
```

Comment:
release 3?

Reply:
This raised an error after changing flow_controller.rb:79 to raise per your comment above, thanks.